// @jackboxgames/ecast/client.js
class ClientWelcome {
    constructor(options) {
        this.id = options.id;
        this.deviceId = options.deviceId;
        this.name = options.name;
        this.secret = options.secret;
        this.reconnect = options.reconnect;
        this.entities = options.entities;
        this.here = options.here;
        this.profile = options.profile;
        this.replayEnd =  options.replayEnd;
    }
}

class ClientConnected {
    constructor(options) {
        this.id = options.id;
        this.userId = options.userId;
        this.name = options.name;
        this.role = options.role;
        this.reconnect = options.reconnect;
    }
}

class ClientDisconnected {
    constructor(options) {
        this.id = options.id;
        this.role = options.role;
    }
}

class ClientSend {
    constructor(options) {
        this.to = options.to;
        this.from = options.from;
        this.body = options.body;
        this.userId = options.userId;
    }
}

class ClientKicked {
    constructor(options) {
        this.id = options.id;
        this.banned = options.banned;
        this.reason = options.reason;
    }
}

module.exports = {
    ClientConnected,
    ClientDisconnected,
    ClientKicked,
    ClientSend,
    ClientWelcome,
};
